Rewrote the Discovery implementation to be more flexible

webmozart opened the pull request:
Fixes puli/issues#114 (read this ticket first for background information).

This PR decouples the Discovery completely from the resource repository. Resources are only one type of "artifact" that can be bound to a binding type. Another type of artifact supported by default is class names.

Binding types don't have simple string names anymore, but fully-qualified PHP identifiers (class names). These class names may or may not correspond to existing classes/interfaces. Creating actual interfaces makes the use of binding types easier as of PHP 5.5, as then we can use the `::class` constant.

Example with class names:

```php
// in the consumer package
$discovery->addBindingType(new BindingType(Plugin::class));

// in the provider packages
$discovery->addBinding(new ClassBinding(PaymentPlugin::class, Plugin::class));
$discovery->addBinding(new ClassBinding(SecurityPlugin::class, Plugin::class));
$discovery->addBinding(new ClassBinding(AdPlugin::class, Plugin::class));

// in the consumer package
foreach ($discovery->findBindings(Plugin::class) as $binding) {
    $className = $binding->getClassName();
    $plugin = new $className();

    // ...
}
```

Example with resources:

```php
// in the consumer package
$discovery->addBindingType(new BindingType(MessageCatalog::class));

// in the provider packages
$discovery->addBinding(new ResourceBinding('/app/trans/*.yml', MessageCatalog::class));
$discovery->addBinding(new ResourceBinding('/batman/blog/trans/*.yml', MessageCatalog::class));

// in the consumer package
foreach ($discovery->findBindings(MessageCatalog::class) as $binding) {
    // Iterate the resources matched by the globs
    foreach ($binding->getResources() as $resource) {
        $messages = Yaml::load($resource->getBody());

        // ...
    }
}
```

This PR is finished except for PHPDoc and the CHANGELOG. Please review.

stof commented:
> Binding types don't have simple string names anymore, but fully-qualified PHP identifiers (class names).

Saying this looks weird to me. It would be much more flexible to accept any string as name. class names are strings, and so would be valid, while the opposite is not true. Saying you expect a class name but that this class name does not need to exist is very confusing.

webmozart commented:
@stof We need some way to avoid naming conflicts. Class identifiers naturally support this by adding a vendor namespace, like `Stof\MessageCatalog`.

Now we could add a requirement for such a class to exist. But I know that some people are very sensitive about creating "unnecessary" classes/interfaces, so I don't want to add this requirement.

Accepting any kind of string as type name is too fragile IMO as we can't prevent conflicts.

stof commented:
@webmozart Document that the binding type is an identifier and needs to be unique. But don't document it as a class name if there is no class. This would cause more confusion than good IMO.
I could perfectly use ``stof_message_catalog`` as identifier, which also prevents conflicts. Using a class name is not the only way.

And if you don't enforce the class to exist, you are still effectively accepting any string as name. You just document it in a confusing way.

webmozart commented:
@stof Maybe I shouldn't have mentioned that the class doesn't have to exist. I think in 90% of the cases, you *should* create an interface for the binding type, as binding type is just that: An interface between packages.

The recommended way to configure binding types will be by using annotations on that interface, for example:

```php
/**
 * @BindingType
 * @Accept("ResourceBinding")
 * @Parameters({
 *     @Parameter('translation_domain', default = 'messages')
 * })
interface MessageCatalog
{
}
```

There will be an option for configuring binding types in `puli.json` if you don't want to add annotations to your code. There will also be an option for not having to create the interface in that case.

But for the 90% case I think that the above should be recommended, as it is much easier to use and understand (IMO).

stof commented:
> The recommended way to configure binding types will be by using annotations on that interface, for example

how would you discover the binding types ? will you scan all classes of the package for such annotations ? this would not be efficient (and if you use the doctrine annotation loader, it might break things if some classes are meant to be used only when some optional deps are installed and they have annotations coming from them)

stof commented:
and creating an interface for a binding type looks weird to me though. It is an interface which is not meant to be implemented

webmozart commented:
> how would you discover the binding types?

My idea was to add a decorator for the Discovery that checks whether the type has annotations as soon as a type is used that has not been loaded/configured before.

> It is an interface which is not meant to be implemented

For class bindings you can reuse the actual PHP interface, as you usually want to discover classes implementing a specific interface. For resource bindings you are right, I don't think that should be a problem though.

stof commented:
> For resource bindings you are right, I don't think that should be a problem though.

This is the case where I would find it confusing actually.

For class bindings, I agree that using the interface as binding name makes sense. But this can be documented as a recommended practice for class bindings, without forcing the usage of class name for resource bindings.
Btw, if resource bindings are created with names which don't look like class names, it also make it easier to determine whether the binding type is expecting to be used for classes or for resources (binding a class on a binding type expecting resources would break things generally)

stof commented on test.php:
```diff
@@ -0,0 +1,31 @@
+<?php
```
should this file really be committed ?

webmozart replied:
Thanks! :)

stof commented on tests/KeyValueStoreDiscoveryLoadedTest.php:
```diff
@@ -24,15 +24,23 @@ class KeyValueStoreDiscoveryLoadedTest extends AbstractEditableDiscoveryTest
 
+    /**
+     *
+     */
```
empty phpdoc should be removed

stof commented on tests/Fixtures/SerializingArrayStore.php:
```diff
@@ -0,0 +1,187 @@
+ *
+ * @author Bernhard Schussek <bschussek@gmail.com>
+ */
+class SerializingArrayStore implements SortableStore, CountableStore
```
why not moving this class to the KeyValueStore package directly ?

webmozart replied:
Do you see any benefit of this class outside of a test?

stof replied:
well, it is mostly useful for testing, but this would avoid to have to reimplement it in every library wanting to test things with an in-memory storage behaving closely to persistent storages.

webmozart replied:
I guess you're right. I'll move this over.

webmozart replied:
Added in webmozart/key-value-store@3a6670a1903594d2ff36455a3f8943e9b5abdee2

stof commented on tests/Fixtures/SerializingArrayStore.php:
```diff
@@ -0,0 +1,187 @@
+<?php
+
+/*
+ * This file is part of the %%Project%% package.
```
wrong header

stof replied:
same in all new files btw

webmozart replied:
Fixed, thanks!

webmozart commented:
I'll think about it. I'm worried that if we don't provide guidance for type names, things will get messy very fast. Think about Composer's package names. There is a reason why they are enforced to have a specific format.

> if resource bindings are created with names which don't look like class names, it also make it easier to determine whether the binding type is expecting to be used for classes or for resources

You can check that in the configuration of the type. The type can be restricted with the `@Accept` annotation (or equivalent configuration in `puli.json`) - or not, if you wan to create a type that accepts different kinds of bindings.

You could also add your own bindings, by the way. One example is to use a service locator across packages:

```php
// in the consumer package
$discovery->addBindingType(new BindingType(Plugin::class));

// in the provider packages
$discovery->addBinding(new ServiceBinding('payment_plugin', Plugin::class));
$discovery->addBinding(new ServiceBinding('security_plugin', Plugin::class));
$discovery->addBinding(new ServiceBinding('ad_plugin', Plugin::class));

// bindings are hooked to a ServiceLocator instance by a ServiceBindingInitializer

// in the consumer package
foreach ($discovery->findBindings(Plugin::class) as $binding) {
    $plugin = $binding->getService();

    // ...
}
```

Here - again - using the interface names of the expected services makes sense.

stof commented:
> There is a reason why they are enforced to have a specific format.

the strongest reason is because they translate directly to filesystem path though.

> I'll think about it. I'm worried that if we don't provide guidance for type names, things will get messy very fast.

Defining a best practice saying you should namespace your binding type name is not the same than saying it must be a class name IMO.

webmozart commented:
ping @tgalopin

webmozart commented:
I updated this PR now and finished the PHPDoc/CHANGELOG. I renamed "type classes" to "type names" everywhere and did not restrict type names for now. Let's use this a bit in practice and see whether we can find good rules.

webmozart commented:
@tgalopin @stof Could you check what you think about this now?

stof commented on src/Api/Binding/Binding.php:
```diff
@@ -0,0 +1,112 @@
+    /**
+     * Returns the class name of the bound type.
+     *
+     * @return string The class name of the bound type.
```
you are still talking about class name here

webmozart replied:
fixed

stof commented on src/Binding/AbstractBinding.php:
```diff
@@ -138,39 +181,43 @@ public function hasParameterValue($parameterName)
-        return true;
+    protected function postUnserialize(array $data)
+    {
+        $this->uuid = Uuid::fromString(array_pop($data));
```
passing the array by value in the ``postUnserialize`` function and then changing it by reference here causes a reference mismatch forcing PHP to duplicate the array in memory (the caller still holds a reference to the array in its ``$data`` when the method is called.

There is 2 solutions:

- inline the ``unserialize`` call, removing the ``$data``, which may solve the reference mismatch (not sure how PHP handles releasing the reference to the zval, but it should be OK as there is nothing else than the argument referencing it)
- read values from the array instead of altering the array here (this will avoid the reference mismatch for sure).

Given that you only have 3 elements in the array, it may not be too consuming to duplicate the array here. But this code could be taken as example for other places dealing with bigger arrays, in which case it would become an issue.

stof replied:
inlining the call will not solve the issue when subclasses are overwriting the method though, as they will still have a ``$data`` variable holding a reference while they call the parent method.

webmozart replied:
I changed the method signature to accept `$data` as reference.

stof commented on src/InMemoryDiscovery.php:
```diff
@@ -143,35 +188,96 @@ public function getBindings()
+     */
+    protected function hasBindingsWithTypeName($typeName)
+    {
+        return isset($this->bindingsByTypeName[$typeName]) && count($this->bindingsByTypeName[$typeName]) > 0;
```
this can be simplified:

```php
return !empty($this->bindingsByTypeName[$typeName]);
```

webmozart replied:
fixed

stof commented on src/KeyValueStoreDiscovery.php:
```diff
@@ -37,239 +37,462 @@ class KeyValueStoreDiscovery extends AbstractEditableDiscovery
+            $this->initializeBindings($this->bindingsByKey[$key]);
+        }
+
+        foreach ($this->bindingsByKey[$key] as $i => $binding) {
```
``$i`` is unused

webmozart replied:
fixed

webmozart commented:
Updated

stof commented on src/Api/Type/BindingType.php:
```diff
@@ -140,4 +169,29 @@ public function getParameterValue($name)
+     */
+    public function acceptsBinding($className)
+    {
+        return empty($this->acceptedBindings) || in_array($className, $this->acceptedBindings, true);
```
this implementation will not play well with inheritance. Is it intended ?

stof commented:
Apart from my comment, it looks good.

@GrahamCampbell the configuration error reported by StyleCI is confusing, and does not play well with cases where a preset changes

stof commented:
there are also a few places where the phpdoc should be improved. See the Scrutinizer report

GrahamCampbell commented:
Regarding the config stuff, you we sent you an email about this 2 weeks ago, and at the weekend. I also wrote a blog post about it. You might want to check that out. :)

stof commented:
@GrahamCampbell I'm not the one receiving emails for Puli though.

GrahamCampbell commented:
I sent the email out to everyone with a StyleCI account.

stof commented:
@GrahamCampbell I don't have one

GrahamCampbell commented:
Sure. We still blogged about it and tweeted it though. There's also a notice posted on our preset docs page.
